pgjdbc/src/main/java/org/postgresql/util/NumberParser.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util;
+
+/**
+ * Optimised byte[] to number parser.
+ */
+public class NumberParser {
+  private static final NumberFormatException FAST_NUMBER_FAILED = new NumberFormatException() {
+    @Override
+    public Throwable fillInStackTrace() {
+      return this;
+    }
+  };
+
+  private static final long MAX_LONG_DIV_TEN = Long.MAX_VALUE / 10;
+
+  /**
+   * Optimised byte[] to number parser. This code does not handle null values, so the caller must do
+   * checkResultSet and handle null values prior to calling this function. Fraction part is discarded.
+   *
+   * @param bytes integer represented as a sequence of ASCII bytes
+   * @return The parsed number.
+   * @throws NumberFormatException If the number is invalid or the out of range for fast parsing.
+   *        The value must then be parsed by another (less optimised) method.
+   */
+  public static long getFastLong(byte[] bytes, long minVal, long maxVal) throws NumberFormatException {
+    int len = bytes.length;
+    if (len == 0 || len > 30) {
+      throw FAST_NUMBER_FAILED;
+    }
+
+    boolean neg = bytes[0] == '-';
+
+    long val = 0;
+    int start = neg ? 1 : 0;
+    while (start < len) {
+      byte b = bytes[start++];
+      if (b < '0' || b > '9') {
+        if (b == '.') {
+          if (neg && len == 2 || !neg && len == 1) {
+            throw FAST_NUMBER_FAILED;
+          }
+          while (start < len) {
+            b = bytes[start++];
+            if (b < '0' || b > '9') {
+              throw FAST_NUMBER_FAILED;
+            }
+          }
+          break;
+        } else {
+          throw FAST_NUMBER_FAILED;
+        }
+      }
+
+      if (val <= MAX_LONG_DIV_TEN) {
+        val *= 10;
+        val += b - '0';
+      } else {
+        throw FAST_NUMBER_FAILED;
+      }
+    }
+
+    if (val < 0) {
```

Comment:
how would val be < 0 ?

Reply:
It is possible to get long overflow in two situations:
1. for incorrect string "9223372036854775809" - so we have to throw exception
2. for MIN_VALUE, "-9223372036854775808" - interesting thing here is that actually we will get correct number thanks to complement arithmetic.

I will add comment.